File: Python_Version/cohort.py
import random, math, copy
from fixed_parameters import *
from human import Human
from klass import Class

class Cohort:

    def __init__(self, num_ex_students, num_ex_teachers, intern_random_infection_rate, extern_random_infection_rate, teacher_break_rules, student_break_rules, houses, class_info):
        self.quarantine = False
        self.warning = False
        self.quarantine_time = 0

        self.crew = [[], []]
        for i in range(num_ex_students):
            spawnpoint = (random.uniform(d, lund_size[0]-d), random.uniform(cwh, lund_size[1]-hwh))
            self.crew[0].append(Human(spawnpoint, student_break_rules, False, extern_random_infection_rate))
        for i in range(num_ex_teachers):
            spawnpoint = (random.uniform(d, lund_size[0]-d), random.uniform(cwh, lund_size[1]-hwh))
            self.crew[1].append(Human(spawnpoint, teacher_break_rules, False, extern_random_infection_rate))

        self.houses = houses
        for i in self.houses:
            self.crew[0].extend(i.crew[0])
            self.crew[1].extend(i.crew[1])

        self.classes = []
        for info in class_info:
            self.classes.append(Class(info[0], info[1]))
            self.classes[-1].pick(self.crew)
            for i in self.classes[-1].crew[0]:
                self.crew[0].remove(i)
            for i in self.classes[-1].crew[1]:
                self.crew[1].remove(i)
        for i in self.classes:
            self.crew[0].extend(i.crew[0])
            self.crew[1].extend(i.crew[1])

    def house_time(self, fakewalls):
        for i in self.houses:
            i.house_time(fakewalls)
        for i in self.crew:
            for j in i:
                if not j.intern:
                    j.position = (None,None)

    def lessoning(self, walls, fakewalls):
        for i in self.crew:
            for j in i:
                if not j.intern:
                    j.position = (random.uniform(d, lund_size[0]-d), random.uniform(cwh, lund_size[1]-hwh))
        for i,c in enumerate(self.classes):
            c.lessoning(walls[i*2:i*2+1], fakewalls)

    def set_Quarantine(self):
        self.quarantine = True
        for i in self.houses:
            i.quarantine = True

    def warn(self):
        self.warning = True
        for i in self.houses:
            i.warning = True
        for i in self.classes:
            i.warning = True

File: Julia_Simulation/cohort.jl
module cohort
    using Random
    include("./fixed_parameters.jl")
    using .fixed_parameters
    include("./human.jl")
    using .human
    include("./class.jl")
    using .class

    mutable struct Cohort
        quarantine::Bool
        warning::Bool
        quarantine_time::Int64
        crew::Array
        houses::Array
        classes::Array

        house_time::Any
        set_Quarantine::Any
        warn::Any

        function Cohort(num_ex_students::Int64, num_ex_teachers::Int64, intern_random_infection_rate::Float64, extern_random_infection_rate::Float64, teacher_break_rules::Float64, student_break_rules::Float64, houses::Array, class_info::Array)
            my_crew = [Array{Any}(undef, num_ex_students), Array{Any}(undef, num_ex_teachers)]
            for i in 1:num_ex_students
                spawnpoint = (rand(d:0.01:lund_size[1]-d), rand(cwh:0.01:lund_size[2]-hwh))
                my_crew[1][i] = Human(spawnpoint, student_break_rules, false, extern_random_infection_rate)
            end
            for i in 1:num_ex_teachers
                spawnpoint = (rand(d:0.01:lund_size[1]-d), rand(cwh:0.01:lund_size[2]-hwh))
                my_crew[2][i] = Human(spawnpoint, teacher_break_rules, false, extern_random_infection_rate)
            end

            for i in houses
                for j in 1:2
                    for k in i.crew[j]
                        push!(my_crew[j], k)
                    end
                end
            end

            my_classes = Array{Any}(undef, length(class_info))
            Random.shuffle!(my_crew[1])
            Random.shuffle!(my_crew[2])
            start = [0,0]
            for (index, info) in enumerate(class_info)
                my_classes[index] = Class(info[1], info[2])
                my_classes[index].pick(my_classes[index], my_crew, start)
                start[1] += my_classes[index].num_students
                start[2] += my_classes[index].num_teachers
            end

            new(false, false, 0, my_crew, houses, my_classes, house_time, set_Quarantine, warn)
        end
    end

    function house_time(self::Cohort, fakewalls::Array)
        for i in self.houses
            i.house_time(i, fakewalls)
        end
        for i in self.crew
            for j in i
                if !j.intern
                    j.position = (nothing, nothing)
                end
            end
        end
    end

    function set_Quarantine(self::Cohort)
        self.quarantine = true
        for i in self.houses
            i.quarantine = true
        end
    end

    function warn(self::Cohort)
        self.warning = true
        for i in self.houses
            i.warning = true
        end
        for i in self.classes
            i.warning = true
        end
    end

    export Cohort

end
